pkg/operation/botanist/namespaces.go

```diff
@@ -78,15 +79,81 @@ func (b *Botanist) DeploySeedNamespace(ctx context.Context) error {
+		if zonePinningRequired(b.Shoot.GetInfo(), b.Seed.GetInfo()) && !metav1.HasLabel(namespace.ObjectMeta, v1beta1constants.ShootZonePinning) {
+			metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, "")
+		}
 
-			if (haAnnotationValues == v1beta1constants.ShootAlphaControlPlaneHighAvailabilitySingleZone && gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes)) ||
-				!haIsConfigured {
-				metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, "")
-			}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	b.SeedNamespaceObject = namespace
+	return nil
+}
+
+func zonePinningRequired(shoot *gardencorev1beta1.Shoot, seed *gardencorev1beta1.Seed) bool {
+	if !gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
+		return false
+	}
+
+	if _, ok := seed.Labels[v1beta1constants.LabelSeedMultiZonal]; !ok {
+		return false
+	}
+
+	haAnnotationValues, ok := shoot.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
+	return !ok || haAnnotationValues == v1beta1constants.ShootAlphaControlPlaneHighAvailabilitySingleZone
+}
+
+// AddZoneInformationToSeedNamespace adds the name of the availability zone
+// in which pods of a non-HA or single-zonal shoot run to the zone-pinning annotation.
+func (b *Botanist) AddZoneInformationToSeedNamespace(ctx context.Context) error {
+	if !zonePinningRequired(b.Shoot.GetInfo(), b.Seed.GetInfo()) {
+		return nil
+	}
+
+	// Let's assume we can take any pod from the list to extract the zone information because they are all scheduled with
+	// a zone affinity added by the pod-zone-affinity webhook of GRM.
+	pods := &corev1.PodList{}
+	if err := b.K8sSeedClient.Client().List(ctx, pods, client.InNamespace(b.Shoot.SeedNamespace)); err != nil {
+		return nil
+	}
+
+	if len(pods.Items) == 0 {
+		return fmt.Errorf("zone information cannot be extracted because no running pods found in control-plane")
+	}
+
+	var nodeName string
+	for _, pod := range pods.Items {
+		if pod.Spec.NodeName == "" {
+			continue
 		}
+		nodeName = pod.Spec.NodeName
+		break
+	}
+
+	if nodeName == "" {
+		return fmt.Errorf("zone information cannot be extracted because no pods have been scheduled yet")
+	}
+
+	node := &corev1.Node{}
+	if err := b.K8sSeedClient.Client().Get(ctx, kutil.Key(nodeName), node); err != nil {
+		return fmt.Errorf("zone information cannot be extracted: %w", err)
+	}
 
+	zone := node.Labels[corev1.LabelTopologyZone]
+	if zone == "" {
+		return fmt.Errorf("zone information cannot be extracted because node %q does not contain any zone information", node.Name)
+	}
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: b.Shoot.SeedNamespace,
+		},
+	}
+
+	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, b.K8sSeedClient.Client(), namespace, func() error {
+		metav1.SetMetaDataLabel(&namespace.ObjectMeta, v1beta1constants.ShootZonePinning, zone)
```

Comment:
Use https://github.com/gardener/gardener/blob/9044fe3cd5c7c59871bc4de9aca914686f0dc4d3/pkg/operation/types.go#L78

Do not call `GetAndCreateOrMergePatch` but rather `Patch` instead (the namespace should already exist).